applications/nrf_desktop/src/hw_interface/selector_hw.c

```diff
@@ -0,0 +1,354 @@
+{
+	u32_t value;
+	const struct pin *sel_pins = pins[selector->id];
+
+	for (size_t i = 0; i < selector->pin_cnt; i++) {
+		int err = gpio_pin_read(gpio_dev[sel_pins[i].port],
+					sel_pins[i].pin,
+					&value);
+
+		if (err) {
+			LOG_ERR("Cannot read value from port:%" PRIu8
+				" pin: %" PRIu8, sel_pins[i].port,
+				sel_pins[i].pin);
+			return err;
+		}
+
+		if (value) {
+			selector->state = i;
+			selector_event_send(selector);
+			return 0;
+		}
+	}
+
+	selector->state = DEFAULT_STATE;
+	LOG_WRN("Selector %" PRIu8 " state is unknown.", selector->id);
+	selector_event_send(selector);
+
+	return 0;
+}
+
+static void selector_isr(struct device *dev, struct gpio_callback *cb,
+			 u32_t pins_mask)
+{
+	struct selector *selector;
+	const struct pin *sel_pins;
+	u8_t port;
+
+	for (size_t i = 0; i < ARRAY_SIZE(selectors); i++) {
+		for (size_t j = 0; j < ARRAY_SIZE(port_map); j++) {
+			if (cb == &selectors[i].gpio_cb[j]) {
+				selector = &selectors[i];
+				sel_pins = pins[i];
+				port = j;
+				break;
+			}
+		}
+	}
+	__ASSERT_NO_MSG(port < ARRAY_SIZE(port_map));
+
+	for (size_t i = 0; i < selector->pin_cnt; i++) {
+		int err = gpio_pin_disable_callback(gpio_dev[sel_pins[i].port],
+						    sel_pins[i].pin);
+
+		if (err) {
+			LOG_ERR("Cannot disable callback (err %d)", err);
+		}
+
+		if ((sel_pins[i].port == port) &&
+		    (pins_mask & BIT(sel_pins[i].pin))) {
+			/* Only one pin per selector may be set at the time. */
+			__ASSERT_NO_MSG((pins_mask & ~BIT(sel_pins[i].pin)) == 0);
+			selector->state = i;
+		}
+	}
+
+	k_delayed_work_submit(&selectors->work, 0);
+}
+
+static int enable_interrupts(struct selector *selector)
+{
+	const struct pin *sel_pins = pins[selector->id];
+	int key = irq_lock();
+
+	for (size_t i = 0; i < selector->pin_cnt; i++) {
+		if (i == selector->state) {
+			continue;
+		}
+
+		int err = gpio_pin_enable_callback(gpio_dev[sel_pins[i].port],
+						   sel_pins[i].pin);
+		if (err) {
+			LOG_ERR("Cannot enable callback (err %d)", err);
+			return err;
+		}
+	}
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+static int disable_interrupts(struct selector *selector)
+{
+	const struct pin *sel_pins = pins[selector->id];
+	int key = irq_lock();
+
+	for (size_t i = 0; i < selector->pin_cnt; i++) {
+		int err = gpio_pin_disable_callback(gpio_dev[sel_pins[i].port],
+						    sel_pins[i].pin);
+
+		if (err) {
+			LOG_ERR("Cannot disable callback (err %d)", err);
+			return err;
+		}
+	}
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+void selector_work_fn(struct k_work *w)
+{
+	struct selector *selector = CONTAINER_OF(w, struct selector, work);
+
+	selector_event_send(selector);
+
+	int err = enable_interrupts(selector);
+
+	if (err) {
+		module_set_state(MODULE_STATE_ERROR);
+	}
+}
+
+static int configure(struct selector *selector)
+{
+	const struct pin *sel_pins = pins[selector->id];
+	u32_t bitmask = 0;
+
+	for (size_t i = 0; i < ARRAY_SIZE(pins[selector->id]); i++) {
+		if ((sel_pins[i].port == 0) && (sel_pins[i].pin == 0)) {
```

Comment:
What is this condition for?

Reply:
Check for first empty record to assign proper length of array records (states number may be different for switches).

Reply:
Where is such "sentinel" in pins array defined in selector_hw_def.h?

Reply:
Where using such initializer for const static array uninitialuzed elements are set to zeros.

Reply:
So CONFIG_DESKTOP_SELECTOR_HW_MAX_STATES must be one greater than number of entries in the pins array entry, it that correct?

Reply:
Could be the same too (there is assignment below).

Reply:
If that is equal to the size then there is no "sentinel" - zeros in array. What if just after end of array linker puts next data?

Reply:
If nothing was assigned selector->pin_cnt equals zero. There is assignment below for loop.